### PR TITLE
Update http authentication docs with new operation level auth and scopes

### PR DESCRIPTION
**Pull request by timotheeguerin**
PR added support for it but we need to update the docs to explain the new feature.

### CODE REVIEW

File: .chronus/changes/docs-operation-level-2024-1-27-19-50-39.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+# Change versionKind to one of: internal, fix, dependencies, feature, deprecation, breaking
+changeKind: internal
+packages:
+  - "@typespec/http"
+  - "@typespec/openapi"
```
Comment (timotheeguerin) [inline .chronus/changes/docs-operation-level-2024-1-27-19-50-39.md:6]:
```suggestion
  - "@typespec/openapi3"
```

File: packages/http/src/types.ts [hunk 2]
```diff
@@ -18,7 +18,7 @@ export type HttpVerb = "get" | "put" | "post" | "patch" | "delete" | "head";
 
 /** @deprecated use Authentication */
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
-type ServiceAuthentication = Authentication;
+export type ServiceAuthentication = Authentication;
```
Comment (timotheeguerin) [inline packages/http/src/types.ts:21]: this was meant to be exported to make this a non breaking change

File: docs/libraries/http/authentication.md [hunk 3]
```diff
@@ -4,40 +4,40 @@ title: Authentication
 
 # Configure Http Authentication
 
-Notes:
+## Configure
 
-- Authentication right now can ONLY be configured at the service level.
+Authentication can be configured using the `@useAuth` decorator on a service namespace, sub namespace, interface or operation.
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:9]:
```suggestion
Authentication can be configured using the `@useAuth` decorator on a service namespace, sub-namespace, interface or operation.
```

File: docs/libraries/http/authentication.md [hunk 4]
```diff
@@ -4,40 +4,40 @@ title: Authentication
...
 
-## Configure
+The `@useAuth` decorator takes as input the security to apply for all operations contained in the type. If another `@useAuth` was specified on a parent type the new one will override the value of the parent. [See application hierarchy](#application-hierarchy)
 
-Authentication can be configured using the `@useAuth` decorator on the service namespace. The decorator accept a few options:
+The input can be a:
 
-- A security scheme (see options [here](https://github.com/microsoft/typespec/blob/main/packages/http/lib/auth.tsp)). This means this is the security scheme to use to authenticate this service.
+- A single security scheme (see options [here](https://github.com/microsoft/typespec/blob/main/packages/http/lib/auth.tsp)). This means this is the security scheme to use to authenticate this service.
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:15]:
```suggestion
- A single security scheme (see options [here](https://github.com/microsoft/typespec/blob/main/packages/http/lib/auth.tsp)).
```

File: docs/libraries/http/authentication.md [hunk 5]
```diff
@@ -4,40 +4,40 @@ title: Authentication
...
 
-```typespec
-@useAuth(Auth1)
-```
+  ```typespec
+  @useAuth(Auth1)
+  ```
 
 - A tuple of security scheme. This means ALL the different security schemes of the tuple MUST be used together to authenticate this service.
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:21]:
```suggestion
- A tuple of security scheme. This means **all** the different security schemes of the tuple **must** be used together to authenticate this service.
```

File: docs/libraries/http/authentication.md [hunk 6]
```diff
@@ -4,40 +4,40 @@ title: Authentication
...
-```typespec
-// Use BOTH Auth1 or Auth2
-@useAuth([Auth1, Auth2])
-```
+  ```typespec
+  // Use BOTH Auth1 or Auth2
+  @useAuth([Auth1, Auth2])
+  ```
 
 - A union of security scheme. This means EITHER of the security schemes can be used to authenticate this service
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:28]:
```suggestion
- A union of security schemes. This means **one** of the security schemes **must** be used to authenticate this service. 
```

"can be used" made me wonder if they were optional.

File: docs/libraries/http/authentication.md [hunk 7]
```diff
@@ -4,40 +4,40 @@ title: Authentication
...
-```typespec
-// Use EITHER Auth1 or Auth2
-@useAuth(Auth1 | Auth2)
-```
+  ```typespec
+  // Use EITHER Auth1 or Auth2
+  @useAuth(Auth1 | Auth2)
+  ```
 
 - A union of tuple security scheme. This means EITHER of the security groups schemes can be used to authenticate this service
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:35]:
```suggestion
- A union of security schemes and tuples representing security groups. This means **one** of the security groups or security schemes **must** be used to authenticate this service.
```

File: docs/libraries/http/authentication.md [hunk 8]
```diff
@@ -111,3 +111,74 @@ OAuth 2.0 is an authorization protocol that gives an API client limited access t
 OAuth relies on authentication scenarios called flows, which allow the resource owner (user) to share the protected content from the resource server without sharing their credentials.
 For that purpose, an OAuth 2.0 server issues access tokens that the client applications can use to access protected resources on behalf of the resource owner.
 For more information about OAuth 2.0, see oauth.net and RFC 6749.
+
+## Application hierarchy
+
+The `@useAuth` decorator can be used on a service, namespace, sub namespace, interface or operation. The security scheme specified will be applied to all operations contained in the type.
+A child typ with a `@useAuth` will **override** the security scheme of the parent type. If the goal is to append a new scheme you'll need to re-specify the same authentication as the parent as well as the new scheme.
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:118]:
```suggestion
A child type with a `@useAuth` will **override** the security scheme of the parent type. If the goal is to append a new scheme, you must re-specify the security schemes on the parent and add the new scheme.
```

File: docs/libraries/http/authentication.md [hunk 9]
```diff
@@ -111,3 +111,74 @@ OAuth 2.0 is an authorization protocol that gives an API client limited access t
...
+op two(): void;
+```
+
+`one` will use `ApiKey` auth and `two` will use the `Basic` auth.
+
+```typespec
+@useAuth(BasicAuth)
+namespace MyService;
+
+@useAuth(ApiKeyAuth<ApiKeyLocation.query, "api_key">) op one(): void; // Use ApiKey only
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:138]:
```suggestion
@useAuth(ApiKeyAuth<ApiKeyLocation.query, "api_key">)
op one(): void; // Use ApiKey only
```

File: docs/libraries/http/authentication.md [hunk 10]
```diff
@@ -111,3 +111,74 @@ OAuth 2.0 is an authorization protocol that gives an API client limited access t
...
+op two(): void; // Use BasicAuth
+```
+
+`one` will use `ApiKey` or `Basic` auth and `two` will use the `Basic` auth.
+
+```typespec
+@useAuth(BasicAuth)
+namespace MyService;
+
+@useAuth(BasicAuth | ApiKeyAuth<ApiKeyLocation.query, "api_key">) op one(): void; // Use ApiKey only
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:148]:
```suggestion
@useAuth(BasicAuth | ApiKeyAuth<ApiKeyLocation.query, "api_key">)
op one(): void; // Use ApiKey only
```

File: docs/libraries/http/authentication.md [hunk 11]
```diff
@@ -111,3 +111,74 @@ OAuth 2.0 is an authorization protocol that gives an API client limited access t
...
+op two(): void; // Use BasicAuth
+```
+
+## OAuth2 Scopes
+
+The `OAuth2` authentication scheme can have a list of scopes that must be granted to the OAuth2 token to be able to use the various operations.
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:154]:
```suggestion
The `OAuth2` security scheme can have a list of scopes that must be granted to the OAuth2 token to be able to use the operations the security scheme applies to.
```

File: docs/libraries/http/authentication.md [hunk 12]
```diff
@@ -111,3 +111,74 @@ OAuth 2.0 is an authorization protocol that gives an API client limited access t
...
+With the creation of a simple alias template you can define a reusable scheme that can be used to specify different scopes for each operations.
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:155]:
```suggestion
When different operations have different scopes, you will likely want to create a template that allows providing OAuth scopes without having to respecify the other properties of the security scheme:
```

File: docs/libraries/http/authentication.md [hunk 13]
```diff
@@ -111,3 +111,74 @@ OAuth 2.0 is an authorization protocol that gives an API client limited access t
...
+This respect the logic of the `@useAuth` decorator that will override the parent security scheme and scopes.
```
Comment (bterlson) [inline docs/libraries/http/authentication.md:156]:
```suggestion
```

**Review by bterlson — APPROVED** [review]

**Pull request merged**